### PR TITLE
Implement extended Map filters

### LINKED ISSUE #1401: Add new Map methods
What you think about this methods in `Map` interface?
* `filter(BiPredicate<K, V>)`
* `filterKeys(Predicate<K>)`
* `filterValues(Predicate<V>)`
* `removeAll(BiPredicate<K, V>)`
* `removeKeys(Predicate<K>)` / `removeAllKeys(Predicate<K>)`
* `removeValues(Predicate<V>)` / `removeAllValues(Predicate<V>)`

### PR DESCRIPTION
**Pull request by valery1707**
Fix #1401

### CODE REVIEW

File: javaslang/src/main/java/javaslang/collection/AbstractMap.java [hunk 1]
```diff
@@ -126,6 +126,42 @@ public M filter(Predicate<? super Tuple2<K, V>> predicate) {
...
+    }
+
+    @Override
+    public M filterValues(Predicate<? super V> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        return filter(t -> predicate.test(t._2()));
+    }
+
+    @Override
+    public Map<K, V> removeAll(BiPredicate<? super K, ? super V> predicate) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/AbstractMap.java]: the return type should be `M`

File: javaslang/src/main/java/javaslang/collection/AbstractMap.java [hunk 2]
```diff
@@ -126,6 +126,42 @@ public M filter(Predicate<? super Tuple2<K, V>> predicate) {
...
+        Objects.requireNonNull(predicate, "predicate is null");
+        return filter(predicate.negate());
+    }
+
+    @Override
+    public Map<K, V> removeKeys(Predicate<? super K> predicate) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/AbstractMap.java]: dito

File: javaslang/src/main/java/javaslang/collection/AbstractMap.java [hunk 3]
```diff
@@ -126,6 +126,42 @@ public M filter(Predicate<? super Tuple2<K, V>> predicate) {
...
+        Objects.requireNonNull(predicate, "predicate is null");
+        return filterKeys(predicate.negate());
+    }
+
+    @Override
+    public Map<K, V> removeValues(Predicate<? super V> predicate) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/AbstractMap.java]: dito

**Comment by danieldietrich** [comment]
Travis-CI build currently breaks with:

```bash
[ERROR] /home/travis/build/javaslang/javaslang/javaslang/src/main/java/javaslang/collection/TreeMap.java:[27,14]
javaslang.collection.TreeMap is not abstract and does not override abstract method
removeValues(java.util.function.Predicate<? super V>) in javaslang.collection.SortedMap
```

File: javaslang/src/main/java/javaslang/collection/AbstractMap.java (hunk 1 above)
Comment (valery1707) [inline javaslang/src/main/java/javaslang/collection/AbstractMap.java]: My bad, sorry

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/AbstractMap.java]: That's why we do code review :) Thanks for the great additions!

**Comment by danieldietrich** [comment]
Great, thank you!

**Pull request merged**